File: tracker/src/test/java/com/cyrillrx/tracker/ScheduledTrackerTest.java
package com.cyrillrx.tracker;

import com.cyrillrx.tracker.event.TrackEvent;
import com.cyrillrx.tracker.utils.Utils;
import org.junit.Assert;
import org.junit.Test;

import java.util.List;
import java.util.concurrent.TimeUnit;

/**
 * @author Cyril Leroux
 *         Created on 22/04/16.
 */
public class ScheduledTrackerTest {

    private static final int EVENT_COUNT = 10;
    private static final int INTERVAL = 200;
    private static final TimeUnit UNIT = TimeUnit.MILLISECONDS;

    @Test
    public void testConsistency() {

        System.out.println("Test started");

        final DummyScheduledTracker tracker = new DummyScheduledTracker(INTERVAL, UNIT);

        // Wait until the tracker actually starts tracking
        TestUtils.wait100Millis();

        Assert.assertTrue("Should be empty", tracker.isEmpty());

        // Track one event
        tracker.track(TestUtils.createFakeEvent(TestUtils.EVENT_CATEGORY));

        // Assert emptiness
        Assert.assertTrue("Should be empty", tracker.isEmpty());

        // Wait enough time for the events to be consumed
        Utils.wait(INTERVAL, UNIT);

        // Assert that the event was sent
        Assert.assertEquals("Should be 1", 1, tracker.getEventCount());

        final List<TrackEvent> trackedEvents = tracker.trackedEvents;
        for (TrackEvent event : trackedEvents) {
            Assert.assertEquals("Category " + TestUtils.EVENT_CATEGORY, TestUtils.EVENT_CATEGORY, event.getCategory());
        }
    }

    @Test
    public void testOneByOne() {

        final DummyScheduledTracker tracker = new DummyScheduledTracker(INTERVAL, UNIT);

        // Wait until the tracker actually starts tracking
        TestUtils.wait100Millis();

        Assert.assertTrue("Should be empty", tracker.isEmpty());

        // Track EVENT_COUNT events
        TestUtils.trackEventsOneByOne(tracker, EVENT_COUNT, TestUtils.EVENT_CATEGORY);

        // Assert emptiness
        Assert.assertTrue("Should be empty", tracker.isEmpty());

        // Wait enough time for the events to be consumed
        Utils.wait(INTERVAL, UNIT);

        // Assert that the events were sent
        Assert.assertEquals("Should be " + EVENT_COUNT, EVENT_COUNT, tracker.getEventCount());
    }

    @Test
    public void testBatch() {

        final DummyScheduledTracker tracker = new DummyScheduledTracker(INTERVAL, UNIT);

        // Wait until the tracker actually starts tracking
        TestUtils.wait100Millis();

        Assert.assertTrue("Should be empty", tracker.isEmpty());

        // Track EVENT_COUNT events
        TestUtils.trackEventsBatch(tracker, EVENT_COUNT, TestUtils.EVENT_CATEGORY);

        // Assert emptiness
        Assert.assertTrue("Should be empty", tracker.isEmpty());

        // Wait enough time for the events to be consumed
        Utils.wait(INTERVAL, UNIT);

        // Assert that the events were sent
        Assert.assertEquals("Should be " + EVENT_COUNT, EVENT_COUNT, tracker.getEventCount());
    }

}

File: tracker/lib/src/androidUnitTest/kotlin/com/cyrillrx/tracker/ScheduledTrackerTest.kt
package com.cyrillrx.tracker

import com.cyrillrx.tracker.event.TrackEvent
import com.cyrillrx.tracker.utils.Utils
import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertTrue

/**
 * @author Cyril Leroux
 * Created on 22/04/16.
 */
class ScheduledTrackerTest {
    @Test
    fun testConsistency() {
        println("Test started")

        val tracker = DummyScheduledTracker(INTERVAL, UNIT)

        // Wait until the tracker actually starts tracking
        TestUtils.wait100Millis()

        assertTrue(tracker.isEmpty(), message = "Should be empty")

        // Track one event
        tracker.track(TrackEvent(TestUtils.EVENT_NAME))

        // Assert emptiness
        assertTrue(tracker.isEmpty(), message = "Should be empty")

        // Wait enough time for the events to be consumed
        Utils.wait(INTERVAL, UNIT)

        // Assert that the event was sent
        assertEquals(1, tracker.getEventCount(), message = "Should be 1")

        val trackedEvents: List<TrackEvent> = tracker.trackedEvents
        for (event in trackedEvents) {
            assertEquals(TestUtils.EVENT_NAME, event.name)
        }
    }

    @Test
    fun testOneByOne() {
        val tracker = DummyScheduledTracker(INTERVAL, UNIT)

        // Wait until the tracker actually starts tracking
        TestUtils.wait100Millis()

        assertTrue(tracker.isEmpty(), message = "Should be empty")

        // Track EVENT_COUNT events
        TestUtils.trackEventsOneByOne(tracker, EVENT_COUNT, TestUtils.EVENT_NAME)

        // Assert emptiness
        assertTrue(tracker.isEmpty(), message = "Should be empty")

        // Wait enough time for the events to be consumed
        Utils.wait(INTERVAL, UNIT)

        // Assert that the events were sent
        assertEquals(EVENT_COUNT, tracker.getEventCount(), message = "Should be $EVENT_COUNT")
    }

    companion object {
        private const val EVENT_COUNT = 10
        private const val INTERVAL = 200L
        private val UNIT: java.util.concurrent.TimeUnit = java.util.concurrent.TimeUnit.MILLISECONDS
    }
}
